src/app/server/DefaultTermsAndConditionsProvider.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "DefaultTermsAndConditionsProvider.h"
+#include "TermsAndConditionsProvider.h"
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLV.h>
+#include <lib/core/TLVTypes.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <lib/support/SafeInt.h>
+
+namespace {
+constexpr chip::TLV::Tag kSerializationVersionTag            = chip::TLV::ContextTag(1);
+constexpr chip::TLV::Tag kAcceptedAcknowledgementsTag        = chip::TLV::ContextTag(2);
+constexpr chip::TLV::Tag kAcceptedAcknowledgementsVersionTag = chip::TLV::ContextTag(3);
+constexpr uint8_t kSerializationVersion                      = 1;
+
+constexpr size_t kEstimatedTlvBufferSize = chip::TLV::EstimateStructOverhead(sizeof(uint8_t),  // SerializationVersion
+                                                                             sizeof(uint16_t), // AcceptedAcknowledgements
+                                                                             sizeof(uint16_t)  // AcceptedAcknowledgementsVersion
+                                                                             ) *
+    2; // Extra space for rollback compatibility
+} // namespace
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsProvider::Init(
+    chip::PersistentStorageDelegate * const inPersistentStorageDelegate,
+    const chip::Optional<chip::app::TermsAndConditions> & inmRequiredAcknowledgements)
+{
+    VerifyOrReturnError(nullptr != inPersistentStorageDelegate, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mPersistentStorageDelegate = inPersistentStorageDelegate;
+    mRequiredAcknowledgements  = inmRequiredAcknowledgements;
+
+    if (CHIP_NO_ERROR == LoadAcceptance(mLatchedAcceptance))
+    {
+        mTemporalAcceptance.SetValue(mLatchedAcceptance.Value());
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsProvider::CheckAcceptance(const Optional<TermsAndConditions> & inTermsAndConditions,
+                                                                         TermsAndConditionsState & outState) const
+{
+    // No validation checks required if no required terms and conditions
+    if (!mRequiredAcknowledgements.HasValue())
+    {
+        outState = TermsAndConditionsState::OK;
+        return CHIP_NO_ERROR;
```

Comment:
Since this always returns CHIP_NO_ERROR, why not simply return a TermsAndConditionsState rather than requiring a CHIP_ERROR check and then an output variable check?